airbyte-lib/airbyte_lib/_processors.py

```diff
@@ -0,0 +1,302 @@
+
+        return batch_id, batch_handle, None
+
+    @abc.abstractmethod
+    def _write_batch(
+        self,
+        stream_name: str,
+        batch_id: str,
+        record_batch: pa.Table | pa.RecordBatch,
+    ) -> BatchHandle:
+        """Process a single batch.
+
+        Returns a batch handle, such as a path or any other custom reference.
+        """
+
+    def _cleanup_batch(  # noqa: B027  # Intentionally empty, not abstract
+        self,
+        stream_name: str,
+        batch_id: str,
+        batch_handle: BatchHandle,
+    ) -> None:
+        """Clean up the cache.
+
+        This method is called after the given batch has been finalized.
+        """
+        pass  # noqa: PIE790 # Intentional no-op
+
+    def _new_batch_id(self) -> str:
+        """Return a new batch handle."""
+        return str(ulid.ULID())
+
+    def _get_batch_handle(
+        self,
+        stream_name: str,
+        batch_id: str | None = None,  # ULID of the batch
+    ) -> str:
+        """Return a new batch handle.
+
+        By default this is a concatenation of the stream name and batch ID.
+        However, any Python object can be returned, such as a Path object.
+        """
+        batch_id = batch_id or self._new_batch_id()
+        return f"{stream_name}_{batch_id}"
+
+    def _finalize_batches(self, stream_name: str) -> dict[str, BatchHandle]:
+        """Finalize all uncommitted batches.
+
+        Returns a mapping of batch IDs to batch handles, for processed batches.
+
+        This is a generic implementation, which can be overridden.
+        """
+        with self._finalizing_batches(stream_name) as batches_to_finalize:
+            if batches_to_finalize and not self.skip_finalize_step:
+                raise NotImplementedError(
+                    "Caches need to be finalized but no _finalize_batch() method "
+                    f"exists for class {self.__class__.__name__}",
+                )
+
+            return batches_to_finalize
+
+    @final
+    @contextlib.contextmanager
+    def _finalizing_batches(
+        self,
+        stream_name: str,
+    ) -> Generator[dict[str, BatchHandle], str, None]:
+        """Context manager to use for finalizing batches, if applicable.
+
+        Returns a mapping of batch IDs to batch handles, for those processed batches.
+        """
+        batches_to_finalize = self._pending_batches[stream_name].copy()
+        state_messages_to_finalize = self._pending_state_messages[stream_name].copy()
+        self._pending_batches[stream_name].clear()
+        self._pending_state_messages[stream_name].clear()
+        yield batches_to_finalize
+
+        self._finalized_batches[stream_name].update(batches_to_finalize)
+        self._finalized_state_messages[stream_name] += state_messages_to_finalize
+
+        for batch_id, batch_handle in batches_to_finalize.items():
+            self._cleanup_batch(stream_name, batch_id, batch_handle)
```

Comment:
The file writer does have a cleanup method as of now, and it inherits from here. So when this runs in the file writer class, it should (in theory) correctly delete the files. I didn't have time to test but I at least wanted to get the implementation "hooks" in place.

Caches can also implement their own cleanup, if applicable, but they shouldn't have to cleanup files because the file writer class should handle that.

Reply:
I realized the no-op implementation gave a false inference. Expanded docstrings here, to explain this better:

- https://github.com/airbytehq/airbyte/pull/33607/commits/1d6d079a1f75597ca0be0534d5215791cb954668

Reply:
New integration test is added which confirms:

1. Files are written where expected.
2. Files are deleted afterwards.

I introduce "cleanup" as a boolean in the file writer config which makes it easy to test this - and also supports use cases where we prefer not to cleanup the files afterwards.